File: m3core/python/m3/component.py
#M3 -- Meka Robotics Robot Components
#Copyright (c) 2010 Meka Robotics
#Author: edsinger@mekabot.com (Aaron Edsinger)

#M3 is free software: you can redistribute it and/or modify
#it under the terms of the GNU Lesser General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#M3 is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU Lesser General Public License for more details.

#You should have received a copy of the GNU Lesser General Public License
#along with M3.  If not, see <http://www.gnu.org/licenses/>.

import yaml
import os 
import m3.toolbox_core as m3t
import m3.component_base_pb2 as mbs


class M3Component:
    """Base class for all Python M3 components.
    The component's name is always prefixed by its type.
    For example: m3actutaor_ma0_j0: type_limbId_jointId
    This convention is used by the M3RtProxy to negotiate data transfer 
    and to locate the component config file.
    """
    def __init__(self,name,type):
        self.name=name
        self.config_name=m3t.get_component_config_filename(name)
        self.config=None
        self.type=type
        self.status=None
        self.command=None
        self.param=None


    def read_config(self):
        """Assumes that the config file is shared between the 
	server and proxy (NFS, etc)"""
        try:
            f=file(self.config_name,'r')
            self.config= yaml.safe_load(f.read())
        except (IOError, EOFError):
            print 'Config file not present:',self.config_name
            return
        if self.config.has_key('param'):
            self.load_attr_from_config(self.config['param'],self.param)

    def write_config(self):
        """Assumes that the config file is shared between the 
	server and proxy (NFS, etc)"""
        if self.config.has_key('param'):
            self.load_config_from_attr(self.config['param'],self.param)
        try:
            f=file(self.config_name,'w')
            print 'Saving...',self.config_name
            f.write(yaml.safe_dump(self.config, default_flow_style=False,width=200))
            f.close()
        except (IOError, EOFError):
            print 'Config file not present:',self.config_file
            return

    def load_attr_from_config(self,config, obj):
        """Load message fields from config dictionary"""
        for k in config:
            if hasattr(obj,k):
		if type(config[k])==list: #repeated field
                    attr=getattr(obj,k)
                    for i in range(len(config[k])):
                        if i<len(attr):
                            attr[i]=config[k][i]
                        else:
                            attr.append(config[k][i])
		elif type(config[k])==dict: #Nested structure
		    self.load_attr_from_config(config[k],getattr(obj,k))
                else:
                    setattr(obj,k,config[k])

		    
    def load_config_from_attr(self,config,obj):
	"""Load config dictionary from message fields"""
	for k in config:
	    if hasattr(obj,k):
		    a=getattr(obj,k)
		    if  hasattr(a,'__setitem__'): #repeated buf:
			    config[k]=list(a)
		    elif hasattr(a,'__getitem__'): # type google.protobuf.reflection.RepeatedCompositeFieldContainer
			    for i in range(len(a)):
				    self.load_config_from_attr(config[k][i],a[i])        
		    elif hasattr(a,'HasField'): #A protobuf struct
			    self.load_config_from_attr(config[k],a)
		    else:
			    config[k]=a	
					

    def update_status(self):
        """Called after every status msg recvd from proxy. Child may override if special processing required."""
        pass 

    def load_command(self):
        """Called before every command msg sent from proxy. Child may override if special processing required."""
        pass 

    def load_param(self):
        """Called before every param msg sent from proxy. Child may override if special processing required."""
        pass 

    def pretty_print_command(self):
        print self.command
    def pretty_print_status(self):
        print self.status
    def pretty_print_param(self):
        print self.param

        # ############################## Private methods ########################################

    def set_int_array(self,attr,val,ind):
        if ind is None:
            ind=range(len(attr))
        if hasattr(ind,'__getitem__'): #indice list given
            if type(val)==float or type(val)==int:
                for i in ind:
                    attr[i]=int(val)
            else:
                j = 0
                for i in ind:
                    attr[i]=int(val[j])
                    j+=1
        else: #single index given
            attr[int(ind)]=int(val)

    def set_float_array(self,attr,val,ind):
        if ind is None:
            ind=range(len(attr))
        if hasattr(ind,'__getitem__'): #indice list given
            if type(val)==float or type(val)==int:
                for i in ind:
                    attr[i]=float(val)
            else:
                j = 0
                for i in ind:
                    attr[i]=float(val[j])
                    j+=1
        else: #single index given
            attr[int(ind)]=float(val)
            
    

File: m3core/src/m3rt/base/component.cpp
/* 
M3 -- Meka Robotics Real-Time Control System
Copyright (c) 2010 Meka Robotics
Author: edsinger@mekabot.com (Aaron Edsinger)

M3 is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

M3 is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with M3.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <m3rt/base/component.h>

namespace m3rt
{


bool M3Component::ParseCommand(std::string & s)
{
	if (!GetCommand()->ParseFromString(s))
	{
		M3_ERR("Error in ParseCommand for %s\n",GetName().c_str());
		return false;
	}
	if (!GetCommand()->IsInitialized())
	{
		M3_INFO("Command Message not initialized for %s\n",GetName().c_str());
		return false;
	}
	//int n=GetCommand()->ByteSize();
	return true;
}
bool M3Component::ParseParam(std::string & s)
{
	if (!GetParam()->ParseFromString(s))
	{
		M3_ERR("Error in ParseParam for %s\n",GetName().c_str());
		return false;
	}
	if (!GetParam()->IsInitialized())
	{
		M3_INFO("Param Message not initialized for %s\n",GetName().c_str());
		return false;
	}
	//int n=GetParam()->ByteSize();
	return true;
}
bool M3Component::SerializeStatus(std::string & s)
{
	//M3_INFO("Serialize: %s with size %d %d %d\n",GetName().c_str(),s.size(),GetStatus()->ByteSize(),GetBaseStatus()->ByteSize());		
	//FixMe: We get occaisional ByteSize() error in message.cc line 235, protocol buffers, on Serialize to String
	//Fix appears to be to do with incorrect ByteSize in the EC messages (nested)
	//Ec byte size changes strangley depending on ctrl mode...
	int n=GetStatus()->ByteSize();
	int q=GetBaseStatus()->ByteSize();
	if (!GetStatus()->IsInitialized())
	{
		M3_INFO("Status Message not initialized for %s\n",GetName().c_str());
		return false;
	}
	if (!GetStatus()->SerializeToString(&s))
	{
		M3_ERR("Error in SerializeToString for %s\n",GetName().c_str());
		return false;
	}
	return true;
}
	
void M3Component::PrettyPrint()
{
	BannerPrint(80,"M3 Component");
	M3_PRINTF("Name: %s\n",GetName().c_str());
	M3_PRINTF("Version: %s\n",GetBaseStatus()->version().c_str());
	M3_PRINTF("Timestamp (uS): %lld\n",GetBaseStatus()->timestamp());
	const char* m3sys_state_names[]={
		"M3COMP_STATE_INIT",
		"M3COMP_STATE_ERR",
		"M3COMP_STATE_SAFEOP",
		"M3COMP_STATE_OP"};
		M3_PRINTF("State:  %s\n",m3sys_state_names[(int)GetBaseStatus()->state()]);
	BannerPrint(60,"Status");
	M3_PRINTF("%s\n",GetStatus()->DebugString().c_str());
	BannerPrint(60,"Param");
	M3_PRINTF("%s\n",GetParam()->DebugString().c_str());
	BannerPrint(60,"Command");
	M3_PRINTF("%s\n",GetCommand()->DebugString().c_str());
}

bool M3Component::ReadConfig(const char * filename)
{
	YAML::Node doc;
	string name;
	string version;
	GetYamlDoc(filename, doc);
	doc["name"] >> name;
	GetBaseStatus()->set_name(name);
	if (name=="") return false;
	try 
	{
		doc["version"]>>version;
		GetBaseStatus()->set_version(version);
	} catch(YAML::TypedKeyNotFound<string> e) 
	{
		//M3_WARN("Missing version key in config file for component %s. Defaulting to default\n",name.c_str());
		GetBaseStatus()->set_version("default");
	} 
	
	//Ugly solution...
	//Search to find the registered id to the given name
	int found=0;
	for (int i=0;i<version_names.size();i++)
	{
		if (version_names[i]==GetBaseStatus()->version())
		{
			version_id=version_ids[i];
			found=1;
			break;
		}
	}
	if (!found)
	{
		M3_ERR("Component %s was unable to find registered version %s upon loading\n",GetName().c_str(),GetBaseStatus()->version().c_str());
		return false;
	}
	version_ids.clear(); //no longer needed
	version_names.clear();
	return true;
}

}